### ISSUE TITLE
Saving the Image as Tiff

### ISSUE THREAD

**Issue by mattgbiz** [micro-manager/pycro-manager#38]
Hi, I am working on a project that is aiming to automate a neutron tomography process. I have just started 1 learning Python as a non-coder by trade and 2 using pycro-manager to work with the EMCCD camera that we use. I have been able to figure out some of the Bridge commands like setting the exposure time, my follow up question that will probably have a lot of time saved just asking here and not sifting through documentation I can only slightly read:

In pyrcromanager as a whole, how would I go about saving the image that I got from core.snap_image() as a tiff to a pre-specified directory?

Also, after saving the image will I need to do anything like close the previous image to then take another picture and save that? I currently cannot access the camera and am trying to get some high level coding done ahead of time.

Thanks

**Comment by henrypinkard** [comment]
pycromanager save as tiffs by default when using the`Acquisition` class. Everything you need should be in the first two sections here: https://pycro-manager.readthedocs.io/en/latest/acq_intro.html#multi-dimensional-acquisitions. You shouldn't need access to your actual camera for high-level coding, as micro-manager will abstract away most of the relevant details. I recommend you just copy and paste some of those examples and then go from there. If there is something specifically confusing about the documentation, please explain further.

Alternatively, if you want to continue using the core as it sounds like you've been doing, there are external libraries for saving (like [this](https://www.tjelvarolsson.com/blog/saving-16bit-tiff-files-using-python/)). The details of how you save will depend on those libraries.

**Comment by mattgbiz** [comment]
Thank you this was very helpful. Another issue I just have been having is an output that pops up when I take an image. I am unsure where this attribution error is coming from as I am unsure what or where the attribute _closed is. Here is a couple snips of my code and output.

![MM_ClosedError](https://user-images.githubusercontent.com/33110348/83421751-92e53a80-a3f6-11ea-97b9-0e5a18fc5c7b.JPG)
![MM_ClosedOutput](https://user-images.githubusercontent.com/33110348/83421758-94aefe00-a3f6-11ea-8ea8-a5ce5a311924.JPG)

**Comment by henrypinkard** [comment]
Don't reopen the `Bridge` each time you want to snap an image. This is meant to happen only once. Probably want to split your function into an initialization that gets called once and a second one to snap and save images.

**Comment by mattgbiz** [comment]
Thank you for that tip, is there any way to not get the Attribution error after the code is run? I have created a class and made an initialize function and the take image function; however, everytime I snap an image it kicks out that same error. I am now able to take multiple pictures which achieves the functionality I need, but I would like to avoid cluttering the terminal.

**Comment by henrypinkard** [comment]
Could you post your code that triggers the error (not just a screenshot, but the text itself)?

**Comment by mattgbiz** [comment]
Here is the code that I have been using to test this. I pull this entire code into another script to take images in, at the bottom of the script is an example of how I use it.

#This will be my preliminary test for controlling the camera
from pycromanager import Bridge
#other classes I haven't needed yetAcquisition, multi_d_acquisition_events, JavaObjectShadow
import numpy as np
#import matplotlib.pyplot as plt
import time
import tifffile
#Below is what works when from pycromanager import Bridge is uncommented
###Setup###
#establish communication with Magellan

class CameraControl:

    def __init__(self):
        
        self.bridge = Bridge()
        #get object representing micro-manager core
        self.core = self.bridge.get_core()
        print('No Errors, MM ready')
        #JavaObjectShadow.__del__(core)


    def TakeImage(self,exposure,folder,filename,suffix):
        #does stuff with autoshutter
        self.auto_shutter = self.core.get_property('Core','AutoShutter')
        self.core.set_property('Core','AutoShutter',0)
        #use the input exposure time
        self.core.set_exposure(exposure)
        ### Acquiring Images ###
        #The micromanager core exposes several mechanisms for acquiring images. 
        #In order to not interfere with ohter pycromanager functionality,
        #this is the one that should be used
        self.core.snap_image()
        self.tagged_image=self.core.get_tagged_image()
        #Pixels by default come out as a 1D array. We can reshape them into an image
        self.pixels = np.reshape(self.tagged_image.pix, newshape=[self.tagged_image.tags['Height'],self.tagged_image.tags['Width']])
        print(self.pixels.dtype)
        #folder = 'D:/PythonScripts/Images'
        #filename = 'Attempt3.tif'
        filepath = folder + '/' + filename + str(suffix) + '.tif'
        self.returnedName = filename + str(suffix) + '.tif'
        tifffile.imwrite(filepath,self.pixels,photometric='minisblack')
        return self.returnedName

Attempt = CameraControl()
Attempt.TakeImage(15,'D:/PythonScripts/Images','DoIKnowHowToClasses.tif',1)
Attempt.TakeImage(15,'D:/PythonScripts/Images','DoIKnowHowToClasses.tif',2)

**Comment by henrypinkard** [comment]
That's a lot to parse through. Could you please distill this to a [minimal working example](https://en.wikipedia.org/wiki/Minimal_working_example) to demonstrate the error without all the extra complexity?

**Comment by mattgbiz** [comment]
Here is a minimalish working example, you could even exclude the tifffile and then this is pretty much just what I got from one of the tutorials but I wanted to ensure it still saved an image.
from pycromanager import Bridge 
import numpy as np
import tifffile

bridge = Bridge()
core = bridge.get_core()
core.snap_image()
tagged_image = core.get_tagged_image()
pixels = np.reshape(tagged_image.pix, newshape=[tagged_image.tags['Height'],tagged_image.tags['Width']])
tifffile.imwrite('D:/PythonScripts/Images/ex.tif',pixels,photometric='minisblack')

**Comment by henrypinkard** [comment]
Thanks. This example runs fine for me. Maybe this is a versioning thing. Are you running the latest nightly build of both micro-manager and pycro-manger? Which version of Python are you running?

If you still have an issue with the latest version, can you post the error you get with this minimal working example?

**Comment by mattgbiz** [comment]
Ah yes, I had not updated to latest version of micromanager for a bit, thank you!
